### ISSUE TITLE
 CUDA error: no kernel image is available for execution on the device using Quadro P4000

### ISSUE THREAD

**Issue by davidvilanova** [nanoporetech/dorado#215]
Hi,
I´m trying to run dorado using the following command. My GPU card is a quadro P4000 (8Gb).
I´m running on linux. 

I get the following error.....

```
dorado basecaller -b 64 -x cuda:all --min-qscore 7 --emit-fastq /home/david/sources/dorado/bin/dna_r9.4.1_e8_hac@v3.3 pod5


[2023-05-31 20:49:49.805] [info] > Creating basecall pipeline
[2023-05-31 20:49:50.930] [error] CUDA error: no kernel image is available for execution on the device
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.

Exception raised from c10_cuda_check_implementation at ../c10/cuda/CUDAException.cpp:44 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::string) + 0x57 (0x7f31868a34d7 in /home/david/sources/dorado-0.3.0-linux-x64/bin/../lib/libc10.so)
frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::string const&) + 0x64 (0x7f318686d36b in /home/david/sources/dorado-0.3.0-linux-x64/bin/../lib/libc10.so)
frame #2: c10::cuda::c10_cuda_check_implementation(int, char const*, char const*, int, bool) + 0x118 (0x7f318693fb58 in /home/david/sources/dorado-0.3.0-linux-x64/bin/../lib/libc10_cuda.so)
frame #3: at::native::copy_device_to_device(at::TensorIterator&, bool, bool) + 0x91f (0x7f3139d843af in /home/david/sources/dorado-0.3.0-linux-x64/bin/../lib/libtorch_cuda.so)
frame #4: <unknown function> + 0x1985b76 (0x7f3139d85b76 in /home/david/sources/dorado-0.3.0-linux-x64/bin/../lib/libtorch_cuda.so)
frame #5: <unknown function> + 0x1d4b670 (0x7f318874b670 in /home/david/sources/dorado-0.3.0-linux-x64/bin/../lib/libtorch_cpu.so)
frame #6: at::native::copy_(at::Tensor&, at::Tensor const&, bool) + 0x62 (0x7f318874c802 in /home/david/sources/dorado-0.3.0-linux-x64/bin/../lib/libtorch_cpu.so)
frame #7: at::_ops::copy_::redispatch(c10::DispatchKeySet, at::Tensor&, at::Tensor const&, bool) + 0x7b (0x7f318935e4ab in /home/david/sources/dorado-0.3.0-linux-x64/bin/../lib/libtorch_cpu.so)
frame #8: <unknown function> + 0x5442fb5 (0x7f318be42fb5 in /home/david/sources/dorado-0.3.0-linux-x64/bin/../lib/libtorch_cpu.so)
frame #9: at::_ops::copy_::redispatch(c10::DispatchKeySet, at::Tensor&, at::Tensor const&, bool) + 0x7b (0x7f318935e4ab in /home/david/sources/dorado-0.3.0-linux-x64/bin/../lib/libtorch_cpu.so)
frame #10: <unknown function> + 0x5445364 (0x7f318be45364 in /home/david/sources/dorado-0.3.0-linux-x64/bin/../lib/libtorch_cpu.so)
frame #11: at::_ops::copy_::call(at::Tensor&, at::Tensor const&, bool) + 0x15f (0x7f31893bb7af in /home/david/sources/dorado-0.3.0-linux-x64/bin/../lib/libtorch_cpu.so)
frame #12: dorado() [0x58c34b]
frame #13: dorado() [0x4e9e33]
frame #14: dorado() [0x5372d6]
frame #15: dorado() [0x50be13]
frame #16: dorado() [0x50bf20]
frame #17: dorado() [0x5383d0]
frame #18: dorado() [0x50bcbd]
frame #19: dorado() [0x589738]
frame #20: dorado() [0x58a809]
frame #21: dorado() [0x5893b6]
frame #22: dorado() [0x4aa0ad]
frame #23: dorado() [0x4aecec]
frame #24: dorado() [0x468d93]
frame #25: <unknown function> + 0x23850 (0x7f3132439850 in /usr/lib/libc.so.6)
frame #26: __libc_start_main + 0x8a (0x7f313243990a in /usr/lib/libc.so.6)
frame #27: dorado() [0x46e76f]

➜  P31910_N057_E09_DORADO nvidia-smi 
Wed May 31 20:50:04 2023       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 530.41.03              Driver Version: 530.41.03    CUDA Version: 12.1     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                  Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf            Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  Quadro P4000                    On | 00000000:09:00.0 Off |                  N/A |
| 54%   62C    P0               31W / 105W|    444MiB /  8192MiB |      1%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A       987      G   /usr/lib/Xorg                               216MiB |
|    0   N/A  N/A      1803      G   xfwm4                                         2MiB |
|    0   N/A  N/A      1992      C   /usr/NX/bin/nxnode.bin                      115MiB |
|    0   N/A  N/A      2210      G   /usr/lib/insync/insync                       83MiB |
|    0   N/A  N/A   3521370      G   ...64245944,2215385247477191414,131072       20MiB |
+---------------------------------------------------------------------------------------+
➜  P31910_N057_E09_DORADO 
```

**Comment by tijyojwad** [comment]
Hi @davidvilanova - please see details on this related issue 
https://github.com/nanoporetech/dorado/issues/37

TL;DR - this is a Pascal board, but `dorado` only support GPUs with architecture Volta onwards.

**Comment by MarkiyanSV** [comment]
Use dorado 0.2.2 on your pascal card (it can work with all dorado 0.3.0 models):
https://github.com/nanoporetech/dorado/issues/211#issuecomment-1570312529

PS: If you have libkoi 0.3.x sources, than you can add following optrions to config and re compile it: **-gencode=arch=compute_61,code=sm_61 \**

https://arnon.dk/matching-sm-architectures-arch-and-gencode-for-various-nvidia-cards/

**Comment by vellamike** [comment]
This highlights the need for Dorado to terminate gracefully when a user tries to run with an unsupported card.

**Comment by davidvilanova** [comment]
Thanks fot the update !!!

**Comment by tijyojwad** [comment]
> Use dorado 0.2.2 on your pascal card (it can work with all dorado 0.3.0 models):

@MarkiyanSV Just a note - there's bug fix in 0.3.0 which is needed for running the 5khz models (v4.2.0) correctly.

**Comment by davidvilanova** [comment]
Works well with dorado 0.2.2 and 0.3 models with my GPU card. 

Why is this card supported in 0.2.2 and not anymore in 0.3 ?

**Comment by tijyojwad** [comment]
Hi @davidvilanova - we will look into adding back pascal support. we had removed it previously because of some new kernels that needed Volta+ architectures.

**Comment by chlo0423** [comment]
> > Use dorado 0.2.2 on your pascal card (it can work with all dorado 0.3.0 models):
> 
> @MarkiyanSV Just a note - there's bug fix in 0.3.0 which is needed for running the 5khz models (v4.2.0) correctly.

Hi, 

I have just downloaded v0.2.2. as I have pascal and was having the same issue when running 0.4.2 - @tijyojwad does this mean I won't be able to run the v4.2.0 model as I'm looking at 6mA mods and I think this is the only model I can use?

Thanks!

**Comment by HalfPhoton** [comment]
@chlo0423 , I believe so - v4.2.0 models are not supported in dorado 0.2.2 

Closing this ticket as it's very similar to: https://github.com/nanoporetech/dorado/issues/345 
We'll report progress on our decision on which older GPUs to support there.

Kind regards,
Rich
